File: onebusaway-android/src/main/java/org/onebusaway/android/directions/util/CustomAddress.java
/*
 * Copyright 2012 University of South Florida
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.onebusaway.android.directions.util;

import android.location.Address;
import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

import org.geojson.Feature;
import org.geojson.Point;

import java.util.ArrayList;
import java.util.Locale;


/**
 * Created by foucelhas on 18/08/14.
 */
public class CustomAddress extends Address {

    // Seems that subThoroughFare is street number and thoroughfare is street name.

    private static final int ADDRESS_MAX_LINES_TO_SHOW = 5;

    private boolean isTransitCategory = false;

    public CustomAddress(Locale locale) {
        super(locale);
    }

    public CustomAddress() {
        super(Locale.getDefault());
    }

    /**
     * Creates a CustomAddress out of a Android Address from the Android platform Geocoder API
     * @param address
     */
    public CustomAddress(Address address) {
        super(address.getLocale());
        for (int i = 0; i <= address.getMaxAddressLineIndex(); i++) {
            super.setAddressLine(i, address.getAddressLine(i));
        }
        super.setFeatureName(address.getFeatureName());
        super.setAdminArea(address.getAdminArea());
        super.setSubAdminArea(address.getSubAdminArea());
        super.setLocality(address.getLocality());
        super.setSubLocality(address.getSubLocality());
        super.setThoroughfare(address.getThoroughfare());
        super.setSubThoroughfare(address.getSubThoroughfare());
        super.setPostalCode(address.getPostalCode());
        super.setCountryCode(address.getCountryCode());
        super.setCountryName(address.getCountryName());
        super.setLatitude(address.getLatitude());
        super.setLongitude(address.getLongitude());
        super.setPhone(address.getPhone());
        super.setUrl(address.getUrl());
        super.setExtras(address.getExtras());
    }

    /**
     * Creates a CustomAddress out of a GeoJSON Feature (e.g., from Pelias Autocomplete API)
     * @param address
     */
    public CustomAddress(Feature address) {
        super(Locale.getDefault());

        // TODO - see if we need to fill out more fields
        // For example response with fields see https://github.com/CUTR-at-USF/pelias-client-library/blob/master/src/test/resources/autocomplete-with-focus.json
        // and https://github.com/CUTR-at-USF/pelias-client-library/blob/master/src/test/java/edu/usf/cutr/pelias/AutocompleteTest.java#L42
        super.setAddressLine(0, (String) address.getProperties().get("name"));
        super.setFeatureName((String) address.getProperties().get("label"));
//        super.setAdminArea(address.getAdminArea());
//        super.setSubAdminArea(address.getSubAdminArea());
//        super.setLocality((String) address.getProperties().get("locality"));
//        super.setSubLocality(address.getSubLocality());
//        super.setThoroughfare(address.getThoroughfare());
//        super.setSubThoroughfare(address.getSubThoroughfare());
        super.setPostalCode((String) address.getProperties().get("postalcode"));
//        super.setCountryCode(address.getCountryCode());
        super.setCountryName((String) address.getProperties().get("country"));
//        super.setPhone(address.getPhone());
//        super.setUrl(address.getUrl());
//        super.setExtras(address.getExtras());

        Point p = (Point) address.getGeometry();
        super.setLatitude(p.getCoordinates().getLatitude());
        super.setLongitude(p.getCoordinates().getLongitude());

        // Check if the geocoder marked this location as having a public transportation category
        ArrayList<String> categories = address.getProperty("category");
        if (categories != null) {
            for (String category : categories) {
                if (category.equalsIgnoreCase("transport:public")) {
                    isTransitCategory = true;
                    break;
                }
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        if (getFeatureName() != null) {
            sb.append(getFeatureName());
        }
        if (getSubThoroughfare() != null && !getSubThoroughfare().equals(getFeatureName())) {
            sb.append(", " + getSubThoroughfare());
        }
        if (getThoroughfare() != null && !getThoroughfare().equals(getFeatureName())) {
            sb.append(" " + getThoroughfare());
        }
        if (getSubAdminArea() != null && !getSubAdminArea().equals(getFeatureName())) {
            sb.append(", " + getSubAdminArea());
        }
        if (getLocality() != null && !getLocality().equals(getFeatureName())) {
            sb.append(", " + getLocality());
        }
        if (TextUtils.isEmpty(sb.toString())) {
            int maxLines = (ADDRESS_MAX_LINES_TO_SHOW > getMaxAddressLineIndex())
                    ? getMaxAddressLineIndex() + 1 : ADDRESS_MAX_LINES_TO_SHOW;
            sb.append(getAddressLine(0));
            for (int i = 1; i < maxLines; i++) {
                if (getAddressLine(i) != null) {
                    sb.append(", " + getAddressLine(i));
                }
            }
        }
        return sb.toString();
    }

    public static final Parcelable.Creator<CustomAddress> CREATOR =
            new Parcelable.Creator<CustomAddress>() {
                public CustomAddress createFromParcel(Parcel in) {
                    String language = in.readString();
                    String country = in.readString();
                    Locale locale = country.length() > 0 ?
                            new Locale(language, country) :
                            new Locale(language);
                    CustomAddress a = new CustomAddress(locale);

                    int N = in.readInt();
                    if (N > 0) {
                        for (int i = 0; i < N; i++) {
                            int index = in.readInt();
                            String line = in.readString();
                            a.setAddressLine(index, line);
                        }
                    }

                    a.setFeatureName(in.readString());
                    a.setAdminArea(in.readString());
                    a.setSubAdminArea(in.readString());
                    a.setLocality(in.readString());
                    a.setSubLocality(in.readString());
                    a.setThoroughfare(in.readString());
                    a.setSubThoroughfare(in.readString());
                    a.setPremises(in.readString());
                    a.setPostalCode(in.readString());
                    a.setCountryCode(in.readString());
                    a.setCountryName(in.readString());
                    boolean mHasLatitude = in.readInt() != 0;
                    if (mHasLatitude) {
                        a.setLatitude(in.readDouble());
                    }
                    boolean mHasLongitude = in.readInt() != 0;
                    if (mHasLongitude) {
                        a.setLongitude(in.readDouble());
                    }
                    a.setPhone(in.readString());
                    a.setUrl(in.readString());
                    a.setExtras(in.readBundle());
                    return a;
                }

                public CustomAddress[] newArray(int size) {
                    return new CustomAddress[size];
                }
            };

    /**
     * Is this custom address set.
     *
     * @return true if this address has a valid latitude and longitude
     */
    public boolean isSet() {
        return getLatitude() != Double.MAX_VALUE && getLongitude() != Double.MAX_VALUE;
    }

    /**
     * Create a blank CustomAddress.
     *
     * @return CustomAddress with default locale and unset latitude and longitude.
     */
    public static CustomAddress getEmptyAddress() {
        Locale locale = Locale.getDefault();
        CustomAddress addr = new CustomAddress(locale);
        addr.setLatitude(Double.MAX_VALUE);
        addr.setLongitude(Double.MAX_VALUE);
        return addr;
    }

    /**
     * Return true if this location has been labeled under the category of "public transportation",
     * false if it has not
     *
     * @return true if this location has been labeled under the category of "public transportation",
     * false if it has not
     */
    public boolean isTransitCategory() {
        return isTransitCategory;
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/directions/util/CustomAddress.kt
/*
 * Copyright 2012 University of South Florida
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.directions.util

import android.location.Address
import android.os.Parcel
import android.os.Parcelable
import org.geojson.Feature
import org.geojson.Point
import java.util.Locale

/**
 * Created by foucelhas on 18/08/14.
 */
// Seems that subThoroughFare is street number and thoroughfare is street name.
class CustomAddress(locale: Locale) : Address(locale) {

    /**
     * Return true if this location has been labeled under the category of "public transportation",
     * false if it has not
     */
    var isTransitCategory: Boolean = false
        private set

    constructor() : this(Locale.getDefault())

    /**
     * Creates a CustomAddress out of a Android Address from the Android platform Geocoder API
     */
    constructor(address: Address) : this(address.locale) {
        for (i in 0..address.maxAddressLineIndex) {
            setAddressLine(i, address.getAddressLine(i))
        }
        featureName = address.featureName
        adminArea = address.adminArea
        subAdminArea = address.subAdminArea
        locality = address.locality
        subLocality = address.subLocality
        thoroughfare = address.thoroughfare
        subThoroughfare = address.subThoroughfare
        postalCode = address.postalCode
        countryCode = address.countryCode
        countryName = address.countryName
        latitude = address.latitude
        longitude = address.longitude
        phone = address.phone
        url = address.url
        extras = address.extras
    }

    /**
     * Creates a CustomAddress out of a GeoJSON Feature (e.g., from Pelias Autocomplete API)
     */
    constructor(address: Feature) : this(Locale.getDefault()) {
        // TODO - see if we need to fill out more fields
        // For example response with fields see https://github.com/CUTR-at-USF/pelias-client-library/blob/master/src/test/resources/autocomplete-with-focus.json
        // and https://github.com/CUTR-at-USF/pelias-client-library/blob/master/src/test/java/edu/usf/cutr/pelias/AutocompleteTest.java#L42
        setAddressLine(0, address.properties["name"] as String?)
        featureName = address.properties["label"] as String?
        postalCode = address.properties["postalcode"] as String?
        countryName = address.properties["country"] as String?

        val p = address.geometry as Point
        latitude = p.coordinates.latitude
        longitude = p.coordinates.longitude

        // Check if the geocoder marked this location as having a public transportation category
        val categories: ArrayList<String>? = address.getProperty("category")
        isTransitCategory =
            categories?.any { it.equals("transport:public", ignoreCase = true) } == true
    }

    override fun toString(): String {
        val sb = StringBuilder()
        if (featureName != null) {
            sb.append(featureName)
        }
        if (subThoroughfare != null && subThoroughfare != featureName) {
            sb.append(", $subThoroughfare")
        }
        if (thoroughfare != null && thoroughfare != featureName) {
            sb.append(" $thoroughfare")
        }
        if (subAdminArea != null && subAdminArea != featureName) {
            sb.append(", $subAdminArea")
        }
        if (locality != null && locality != featureName) {
            sb.append(", $locality")
        }
        if (sb.isEmpty()) {
            val maxLines = if (ADDRESS_MAX_LINES_TO_SHOW > maxAddressLineIndex) {
                maxAddressLineIndex + 1
            } else {
                ADDRESS_MAX_LINES_TO_SHOW
            }
            sb.append(getAddressLine(0))
            for (i in 1 until maxLines) {
                if (getAddressLine(i) != null) {
                    sb.append(", ${getAddressLine(i)}")
                }
            }
        }
        return sb.toString()
    }

    /**
     * Is this custom address set.
     *
     * @return true if this address has a valid latitude and longitude
     */
    val isSet: Boolean
        get() = latitude != Double.MAX_VALUE && longitude != Double.MAX_VALUE

    companion object {

        private const val ADDRESS_MAX_LINES_TO_SHOW = 5

        @JvmField
        val CREATOR: Parcelable.Creator<CustomAddress> =
            object : Parcelable.Creator<CustomAddress> {
                override fun createFromParcel(parcel: Parcel): CustomAddress {
                    val language = parcel.readString()
                    val country = parcel.readString()
                    val locale = if (!country.isNullOrEmpty()) {
                        Locale(language, country)
                    } else {
                        Locale(language)
                    }
                    val a = CustomAddress(locale)

                    val n = parcel.readInt()
                    if (n > 0) {
                        for (i in 0 until n) {
                            val index = parcel.readInt()
                            val line = parcel.readString()
                            a.setAddressLine(index, line)
                        }
                    }

                    a.featureName = parcel.readString()
                    a.adminArea = parcel.readString()
                    a.subAdminArea = parcel.readString()
                    a.locality = parcel.readString()
                    a.subLocality = parcel.readString()
                    a.thoroughfare = parcel.readString()
                    a.subThoroughfare = parcel.readString()
                    a.premises = parcel.readString()
                    a.postalCode = parcel.readString()
                    a.countryCode = parcel.readString()
                    a.countryName = parcel.readString()
                    if (parcel.readInt() != 0) {
                        a.latitude = parcel.readDouble()
                    }
                    if (parcel.readInt() != 0) {
                        a.longitude = parcel.readDouble()
                    }
                    a.phone = parcel.readString()
                    a.url = parcel.readString()
                    a.extras = parcel.readBundle()
                    return a
                }

                override fun newArray(size: Int): Array<CustomAddress?> = arrayOfNulls(size)
            }

        /**
         * Create a blank CustomAddress.
         *
         * @return CustomAddress with default locale and unset latitude and longitude.
         */
        @JvmStatic
        fun getEmptyAddress(): CustomAddress {
            val addr = CustomAddress(Locale.getDefault())
            addr.latitude = Double.MAX_VALUE
            addr.longitude = Double.MAX_VALUE
            return addr
        }
    }
}
